src/server/service/debug.rs

```diff
@@ -546,12 +577,220 @@ where
+                    });
+                });
+                // Wait for finishing prepare flashback.
+                tokio::time::timeout(Duration::from_secs(FLASHBACK_TIMEOUT), prepare_wg.wait())
+                    .map_err(|e| {
+                        Error::NotPreparedFlashback(format!(
+                            "prepare flashback timeout err is {}",
+                            e
+                        ))
+                    })
+                    .await?;
+
+                let commit_ts = pd_client
+                    .get_tso()
+                    .await
+                    .expect("failed to get commit_ts from PD");
+                // Flashback to version.
+                let debugger = debugger.lock().await;
+                let flashback_wg = WaitGroup::new();
+                filtered_region_ids.iter().for_each(|&region_id| {
+                    let debugger = debugger.clone();
+                    let work = flashback_wg.clone().work();
+                    runtime.spawn(async move {
+                        let check = debugger
+                            .region_flashback_to_version(region_id, version, start_ts, commit_ts);
+                        match check.await {
+                            Ok(_) => {}
+                            Err(err) => {
+                                return Err(Error::FlashbackFailed(format!(
+                                    "flashback failed err is {}",
+                                    err
+                                )));
+                            }
+                        }
+                        drop(work);
+                        Ok(())
+                    });
+                });
+
+                // Wait for finish.
+                tokio::time::timeout(Duration::from_secs(FLASHBACK_TIMEOUT), flashback_wg.wait())
+                    .map_err(|e| Error::FlashbackFailed(format!("flashback timeout err is {}", e)))
+                    .await?;
+                Ok(FlashbackToVersionResponse::default())
+            })
+            .map(|res| res.unwrap());
+
+        self.handle_response(ctx, sink, f, "debug_flashback_to_version");
+    }
 }
 
+// Check if region's `key_range` intersects with `key_range_limit`.
+pub fn check_intersect_of_range(key_range: &KeyRange, key_range_limit: &KeyRange) -> bool {
+    if !key_range.get_end_key().is_empty()
+        && !key_range_limit.get_start_key().is_empty()
+        && key_range.get_end_key() <= key_range_limit.get_start_key()
+    {
+        return false;
+    }
+    if !key_range_limit.get_end_key().is_empty()
+        && !key_range.get_start_key().is_empty()
+        && key_range_limit.get_end_key() < key_range.get_start_key()
+    {
+        return false;
+    }
+    true
+}
 mod region_size_response {
     pub type Entry = kvproto::debugpb::RegionSizeResponseEntry;
 }
 
 mod list_fail_points_response {
     pub type Entry = kvproto::debugpb::ListFailPointsResponseEntry;
 }
+
+pub struct WaitGroup {
```

Comment:
Can't we use some ready-made implementations of third-party libraries?

Reply:
This is because `crossbeam::WaitGroup` is not async...I did not find a more suitable

Reply:
Well, then I think we should put this struct into a more general place like `tikv_utl`.

Reply:
done.